proto/internal/temporal/server/api/persistence/v1/task_queues.proto

```diff
@@ -0,0 +1,87 @@
+    // (-- api-linter: core::0142::time-field-type=disabled
+    //     aip.dev/not-precedent: Using HLC instead of wall clock. --)
+    temporal.server.api.clock.v1.HybridLogicalClock state_update_timestamp = 3;
+}
+
+// An internal represenation of https://github.com/temporalio/api/blob/9206f1556469515d8e5a82c4d4bf21f2dd9730ce/temporal/api/taskqueue/v1/message.proto#L92
+message CompatibleVersionSet {
+    // Set IDs are used internally by matching.
+    // The first set ID for a given set is the first added build ID.
+    // A set typically has one set ID and extra care is taken to enforce this.
+    // In split brain scenarios, there may be conflicting concurrent writes to the task queue versioning data, in which
+    // case a set might end up with more than one ID.
+    repeated string set_ids = 1;
+    // All the compatible versions, unordered except for the last element, which is considered the set "default".
+    repeated BuildID build_ids = 2;
+    // HLC timestamp representing when the set default was updated. Different from BuildID.state_update_timestamp, which
+    // refers to the build ID status.
+    // (-- api-linter: core::0142::time-field-type=disabled
+    //     aip.dev/not-precedent: Using HLC instead of wall clock. --)
+    temporal.server.api.clock.v1.HybridLogicalClock default_update_timestamp = 3;
+}
+
+// Holds all the data related to worker versioning for a task queue.
+// Backwards-incompatible changes cannot be made, as this would make existing stored data unreadable.
+message VersioningData {
+    // All the incompatible version sets, unordered except for the last element, which is considered the set "default".
+    repeated CompatibleVersionSet version_sets = 1;
+    // HLC timestamp representing when the default set was last updated or established.
+    // (-- api-linter: core::0142::time-field-type=disabled
+    //     aip.dev/not-precedent: Using HLC instead of wall clock. --)
+    temporal.server.api.clock.v1.HybridLogicalClock default_update_timestamp = 2;
+}
+
+// Container for all persistent user provided data for a task queue.
+// Task queue as a named concept here is close to how users interpret them, rather than relating to some specific type
+// (workflow vs activity, etc) and thus, as a consequence, any data that applies to a specific type (say, activity rate
+// limiting) should be defined as such within this structure.
+// This data must all fit in a single DB column and is kept cached in-memory, take extra care to ensure data added here
+// has reasonable size limits imposed on it.
+message TaskQueueUserData {
+    // The last recorded cluster-local Hybrid Logical Clock timestamp.
+    // Updated whenever user data is directly updated due to a user action but not when applying replication events.
+    // The clock is referenced when new timestamps are generated to ensure it produces monotonically increasing
+    // timestamps.
+    temporal.server.api.clock.v1.HybridLogicalClock clock = 1;
```

Comment:
what exactly is this one used for again? I thought the timestamps within versioning_data are enough to merge

Reply:
This keeps the last clock in order to produce a new clock on update.
We need this to ensure that the clock is monotonically increasing in a single cluster context.

Reply:
oh right. when I read reading above I was wondering where it stores the last used timestamp to avoid duplicates, but I didn't connect this to that. to be nitpicky: the comment above should say it's the last recorded cluster-local HLC timestamp _for this task queue_. another task queue might have a later one recorded (due to clock skew) but we never compare them across task queues so it doesn't matter.